viperproject/prusti-dev
# Specifications of foreign functions

Allowing the `extern_spec` attribute on `extern` blocks. Such as

```rust
extern "C" {
    fn mod9(a: i32) -> i32;
}

#[extern_spec(crate)]
extern "C" {
    #[ensures(result >= 0 && result < 9)]
    fn mod9(a: i32) -> i32;
}
```

To test this, adding:
* a `parse/ui` test to check the behavior when the specs don't type check,
* a `cargo_verify` test to test the full compilation of a crate with `#[extern_spec(crate)]` on `extern "C"` and also including a C source file with the implementation of the foreign functions, and
* a `parse/fail` test to test that `#[extern_spec]` (without the path argument) doesn't bind to free functions in the crate root
